gnovm/pkg/gnolang/store.go

```diff
@@ -11,8 +12,11 @@ import (
 	"github.com/gnolang/gno/tm2/pkg/store"
 )
 
-// return nil if package doesn't exist.
-type PackageGetter func(pkgPath string) (*PackageNode, *PackageValue)
+// PackageGetter specifies how the store may retrieve packages which are not
+// already in its cache. PackageGetter should return nil when the requested
+// package does not exist. store should be used to run the machine, or otherwise
+// call any methods which may call store.GetPackage, to avoid import cycles.
+type PackageGetter func(pkgPath string, store Store) (*PackageNode, *PackageValue)
```

Comment:
Why do we need to change this function's signature and pass in a store, while we have `func (ds *defaultStore) SetPackageGetter(pg PackageGetter)` ?

Reply:
The ultimate need is to make sure we have a mechanism, when calling GetPackage, to detect import cycles.

The current implementation uses a `current` map to do such a thing. This is problematic; if two concurrent goroutines are trying to use GetPackage at the same time, they can race and detect import cycles when there are none, or worse not detect them when there are.

To avoid this, we have to pass down a list of the previously imported packages when calling PackageGetter. Doing this as a simple `[]string` with the previous paths would have required this being an additional argument in the entire chain where imports are made; ie. all the usages of predefineNow -> `predefineNow` -> `predefineNow2` -> `tryPredefine`.

It somehow seemed saner to me that import cycles were detected by passing in a different store which keeps track of all of the imports being added up to that point, while still not affecting or changing the baseStore. So, the store parameter serves as the store to be used when further calling the machine by the PackageGetter, which may very well call the machine again and incur in other imports.